<div class="alert alert-info" role="alert">
   
В нашем распоряжении информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.
    
<b>Наша задача — проанализировать базу данных</b>    
    
<b>Описание данных:</b>
    
**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<b>Шаг 1. Загрузим данные и изучим общую информацию </b>

<b>Шаг 2. Посчитаем, сколько книг вышло после 1 января 2000 года</b>
        
<b>Шаг 3. Для каждой книги посчитаем количество обзоров и среднюю оценку</b>    
   
<b>Шаг 4. Определим издательство, которое выпустило наибольшее число книг толще 50 страниц</b>    
        
<b>Шаг 5. Определим автора с самой высокой средней оценкой книг</b>   
       
<b>Шаг 6. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок</b>       
    
<b>Шаг 7. Сформулируем выводы</b>      

    
</div>

# Шаг 1. Загрузим данные и изучим общую информацию

**Подключимся к базе данных**

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

**Посмотрим на данные**

In [2]:
BOOK = ''' 
SELECT  
    *
FROM books
    
        '''
pd.io.sql.read_sql(BOOK, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В данных 'books' у нас информация о 1000 книг

In [3]:
AUTOR = ''' 
SELECT *
FROM authors
        '''
pd.io.sql.read_sql(AUTOR, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В данных 'authors' у нас информация о 636 авторе

In [4]:
PUBLISH = ''' 
SELECT  
    *
FROM publishers
    
        '''
pd.io.sql.read_sql(PUBLISH, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

В данных 'publishers' у нас информация о 340 издательствах

In [5]:
RATING = ''' 
SELECT  
    *
FROM ratings
    
        '''
pd.io.sql.read_sql(RATING, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

В данных 'ratings' у нас информация о пользователях оставивших 6456 оценок о книгах

In [6]:
REVIEW = ''' 
SELECT  
    *
FROM reviews
    
        '''
pd.io.sql.read_sql(REVIEW, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В данных 'reviews' о 2793 обзоров на книги от пользователей

**Вывод по Шагу 1**

Изучив данные из 5 датасетов, мы располагаем следующей информацией:
1000 книг бала написана 636 авторами, которые печатались в 340 издательствах. Данным книгам были оставлены 6456 отзыва и написаны 2793 обзора.

# Шаг 2. Посчитаем, сколько книг вышло после 1 января 2000 года

In [7]:
data = ''' 
SELECT 
    COUNT(title) AS title_count
FROM books
WHERE
    publication_date::date >= '2000-01-01'
        '''
pd.io.sql.read_sql(data, con = engine)

title_count
0          821

**С января 2000 года было выпущено 821 книга**

# Шаг 3. Для каждой книги посчитаем количество обзоров и среднюю оценку

In [8]:
RRA = ''' 
SELECT
    books.book_id,
    subq.review_count,
    subw.mean_rating
FROM
    books
    LEFT JOIN (
        SELECT  
            reviews.book_id,
            COUNT(reviews.review_id) AS review_count
        FROM reviews
        GROUP BY
            reviews.book_id
) AS subq ON subq.book_id = books.book_id 
    LEFT JOIN (
        SELECT  
            ratings.book_id,
            AVG(ratings.rating) AS mean_rating
        FROM ratings
        GROUP BY
            ratings.book_id
) AS subw ON subw.book_id = books.book_id 
ORDER BY
    books.book_id
        '''
pd.io.sql.read_sql(RRA, con = engine)

book_id  review_count  mean_rating
0          1           2.0     3.666667
1          2           1.0     2.500000
2          3           3.0     4.666667
3          4           2.0     4.500000
4          5           4.0     4.000000
..       ...           ...          ...
995      996           3.0     3.666667
996      997           3.0     3.400000
997      998           4.0     3.200000
998      999           2.0     4.500000
999     1000           4.0     3.833333

[1000 rows x 3 columns]

**Мы получили таблицу с книгами, количеством обзоров на них и средней оченкой**

# Шаг 4. Определим издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
BP = ''' 
SELECT  
    publisher,
    COUNT(book_id) AS book_count
FROM books
    LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id 
WHERE 
    num_pages > 50
GROUP BY
    publisher
ORDER BY
    book_count DESC
        '''
pd.io.sql.read_sql(BP, con = engine)

publisher  book_count
0                                        Penguin Books          42
1                                              Vintage          31
2                             Grand Central Publishing          25
3                                     Penguin Classics          24
4                                     Ballantine Books          19
..                                                 ...         ...
329                                         Turtleback           1
330  Atheneum Books for Young Readers: Richard Jack...           1
331                                     Penguin Signet           1
332                                    Victor Gollancz           1
333                      Harvard Business Review Press           1

[334 rows x 2 columns]

**Мы отбросили книги, число страниц которых составляет 50 и менее, тем самым мы исключили из анализа брошюры. Издательство Penguin Books выпустило наибольшее число книг - 42**	

# Шаг 5. Определим автора с самой высокой средней оценкой книг

In [10]:
BAR = ''' 
SELECT  
     authors.author,
     COUNT(ratings.rating_id) AS rating_count,
     AVG(ratings.rating) AS mean_rating
FROM books
    LEFT JOIN ratings ON ratings.book_id = books.book_id
    JOIN authors ON authors.author_id = books.author_id
GROUP BY
    authors.author
HAVING
    COUNT(ratings.rating_id) >= 50       
ORDER BY
    mean_rating DESC
        '''
pd.io.sql.read_sql(BAR, con = engine)

author  rating_count  \
0                                      Diana Gabaldon            50   
1                          J.K. Rowling/Mary GrandPré           312   
2                                     Agatha Christie            53   
3                   Markus Zusak/Cao Xuân Việt Khương            53   
4                                      J.R.R. Tolkien           166   
5                            Roald Dahl/Quentin Blake            62   
6                                   Louisa May Alcott            54   
7                                        Rick Riordan            84   
8                                       Arthur Golden            56   
9                                        Stephen King           106   
10                                       John Grisham            70   
11                                    William Golding            71   
12                                    Nicholas Sparks           111   
13                                       Jodi Picoult            76   
14                                    Sophie Kinsella            57   
15                                    James Patterson            64   
16                                      J.D. Salinger            98   
17           Paulo Coelho/Alan R. Clarke/Özdemir İnce            57   
18  William Shakespeare/Paul Werstine/Barbara A. M...            66   
19                                          Dan Brown           143   
20                                         Lois Lowry            65   
21            George Orwell/Boris Grabnar/Peter Škerl            74   
22                                    Stephenie Meyer           160   
23                                     John Steinbeck            73   

    mean_rating  
0      4.300000  
1      4.288462  
2      4.283019  
3      4.264151  
4      4.240964  
5      4.209677  
6      4.203704  
7      4.130952  
8      4.107143  
9      4.009434  
10     3.971429  
11     3.901408  
12     3.882883  
13     3.881579  
14     3.877193  
15     3.859375  
16     3.846939  
17     3.789474  
18     3.787879  
19     3.741259  
20     3.738462  
21     3.729730  
22     3.662500  
23     3.643836

**Мы отбросили книги, число оценок которых составляет 50 и более. Автор с самой высокой средней оценкой книг является Diana Gabaldon, на ее книги получено 50 оценок и средний рейтинг их составляет 4.3**

# Шаг 6. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [11]:
RR = ''' 
SELECT
    AVG(SUBQ.text) AS avg_week_transaction
FROM    (
        SELECT  
            ratings.username, 
            COUNT(ratings.book_id) AS book_id,
            COUNT(reviews.text) AS text
        FROM ratings
            LEFT JOIN reviews ON reviews.username = ratings.username AND reviews.book_id = ratings.book_id
        GROUP BY
            ratings.username
        HAVING
            COUNT(ratings.book_id) > 50      
        ORDER BY
            book_id DESC
    ) AS SUBQ
        '''
pd.io.sql.read_sql(RR, con = engine)

avg_week_transaction
0             24.333333

In [12]:
pd.read_sql("""
SELECT avg(count) 
FROM (SELECT username, count(text) 
        FROM reviews re WHERE re.username in (
                    SELECT r.username FROM ratings r 
                    GROUP BY r.username
                    HAVING count(r.rating) > 50)
GROUP BY username) as counts
""", con = engine)

avg
0  24.333333

**Среднее количество обзоров от пользователей, которые поставили больше 50 оценок, составляет 24.3**

# Шаг 7. Сформулируем выводы

<div class="alert alert-info" role="alert">
   
В нашем распоряжении информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.
    
<b>Наша задача — проанализировать базу данных</b>    
    
<b>Описание данных:</b>
    
**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<b>Шаг 1. Загрузим данные и изучим общую информацию </b>
    
Изучив данные из 5 датасетов, мы располагаем следующей информацией: 1000 книг бала написана 636 авторами, которые печатались в 340 издательствах. Данным книгам были оставлены 6456 отзыва и написаны 2793 обзора   
    
<b>Шаг 2. Посчитаем, сколько книг вышло после 1 января 2000 года</b>
    
С января 2000 года было выпущено 821 книга  
        
<b>Шаг 3. Для каждой книги посчитаем количество обзоров и среднюю оценку</b>    
    
Мы получили таблицу с книгами, количеством обзоров на них и средней оченкой
   
<b>Шаг 4. Определим издательство, которое выпустило наибольшее число книг толще 50 страниц</b>    
    
Мы отбросили книги, число страниц которых составляет 50 и менее, тем самым мы исключили из анализа брошюры. Издательство Penguin Books выпустило наибольшее число книг - 42    
        
<b>Шаг 5. Определим автора с самой высокой средней оценкой книг</b>   
    
Мы отбросили книги, число оценок которых составляет 50 и более. Автор с самой высокой средней оценкой книг является Diana Gabaldon, на ее книги получено 50 оценок и средний рейтинг их составляет 4.3    
       
<b>Шаг 6. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок</b>    
    
Среднее количество обзоров от пользователей, которые поставили больше 50 оценок, составляет 24.3    
    
<b>Шаг 7. Сформулируем выводы</b>      

    
</div>